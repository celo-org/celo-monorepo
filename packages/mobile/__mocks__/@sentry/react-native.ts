const Sentry = {
  config: () => ({ install: jest.fn() }),
  setTagsContext: jest.fn(),
  captureException: jest.fn(),
}

module.exports = Sentry
