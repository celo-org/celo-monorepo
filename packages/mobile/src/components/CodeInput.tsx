import Card from '@celo/react-components/components/Card'
import TextInput from '@celo/react-components/components/TextInput.v2'
import Checkmark from '@celo/react-components/icons/Checkmark'
import colors from '@celo/react-components/styles/colors'
import fontStyles from '@celo/react-components/styles/fonts.v2'
import { Shadow, Spacing } from '@celo/react-components/styles/styles.v2'
import React, { useLayoutEffect } from 'react'
import {
  ActivityIndicator,
  LayoutAnimation,
  StyleProp,
  StyleSheet,
  Text,
  View,
  ViewStyle,
} from 'react-native'
import ClipboardAwarePasteButton from 'src/components/ClipboardAwarePasteButton'
import { useClipboard } from 'src/utils/useClipboard'

export enum CodeInputStatus {
  DISABLED, // input disabled
  INPUTTING, // input enabled
  PROCESSING, // is the inputted code being processed
  RECEIVED, // is the inputted code received but not yet confirmed
  ACCEPTED, // has the code been accepted and completed
}

export interface Props {
  label: string
  status: CodeInputStatus
  inputValue: string
  inputPlaceholder: string
  inputPlaceholderWithClipboardContent?: string
  onInputChange: (value: string) => void
  shouldShowClipboard: (value: string) => boolean
  multiline?: boolean
  testID?: string
  style?: StyleProp<ViewStyle>
}

export default function CodeInput({
  label,
  status,
  inputValue,
  inputPlaceholder,
  inputPlaceholderWithClipboardContent,
  onInputChange,
  shouldShowClipboard,
  multiline,
  testID,
  style,
}: Props) {
  const [forceShowingPasteIcon, clipboardContent, getFreshClipboardContent] = useClipboard()

  // LayoutAnimation when switching to/from input
  useLayoutEffect(() => {
    LayoutAnimation.easeInEaseOut()
  }, [status === CodeInputStatus.INPUTTING])

  function shouldShowClipboardInternal() {
    if (forceShowingPasteIcon) {
      return true
    }
    return (
      !inputValue.toLowerCase().startsWith(clipboardContent.toLowerCase()) &&
      shouldShowClipboard(clipboardContent)
    )
  }

  const showInput = status === CodeInputStatus.INPUTTING
  const showSpinner = status === CodeInputStatus.PROCESSING || status === CodeInputStatus.RECEIVED
  const showCheckmark = status === CodeInputStatus.ACCEPTED
  const showStatus = showCheckmark || showSpinner

  return (
    <Card
      rounded={true}
      shadow={showInput ? Shadow.SoftLight : null}
      style={[showInput ? styles.containerActive : styles.container, style]}
    >
      {/* These views cannot be combined as it will cause the shadow to be clipped on iOS */}
      <View style={styles.containRadius}>
        <View style={showInput ? styles.contentActive : styles.content}>
          <View style={styles.innerContent}>
            <Text style={showInput ? styles.labelActive : styles.label}>{label}</Text>
            {showInput ? (
              <TextInput
                value={inputValue}
                placeholder={
                  inputPlaceholderWithClipboardContent && shouldShowClipboardInternal()
                    ? inputPlaceholderWithClipboardContent
                    : inputPlaceholder
                }
                onChangeText={onInputChange}
                multiline={multiline}
                testID={testID}
              />
            ) : (
              <Text style={styles.codeValue} numberOfLines={1}>
                {inputValue || ' '}
              </Text>
            )}
          </View>
          {showStatus && (
            <View style={styles.statusContainer}>
              {showSpinner && <ActivityIndicator size="small" color={colors.greenUI} />}
              {showCheckmark && <Checkmark />}
            </View>
          )}
        </View>
        {showInput && (
          <ClipboardAwarePasteButton
            getClipboardContent={getFreshClipboardContent}
            shouldShow={shouldShowClipboardInternal()}
            onPress={onInputChange}
          />
        )}
      </View>
    </Card>
  )
}

const styles = StyleSheet.create({
  container: {
    padding: 0,
    backgroundColor: 'rgba(103, 99, 86, 0.1)',
  },
  containerActive: {
    padding: 0,
  },
  // Applying overflow 'hidden' to `Card` also hides its shadow
  // that's why we're using a separate container
  containRadius: {
    borderRadius: Spacing.Smallest8,
    overflow: 'hidden',
  },
  content: {
    flex: 1,
    flexDirection: 'row',
    alignItems: 'center',
    padding: Spacing.Regular16,
    paddingVertical: Spacing.Small12,
  },
  contentActive: {
    flexDirection: 'row',
    alignItems: 'center',
    padding: Spacing.Regular16,
    paddingBottom: 4,
  },
  innerContent: {
    flex: 1,
  },
  label: {
    ...fontStyles.label,
    color: colors.onboardingBrownLight,
    opacity: 0.5,
    marginBottom: 4,
  },
  labelActive: {
    ...fontStyles.label,
  },
  codeValue: {
    ...fontStyles.regular,
    color: colors.onboardingBrownLight,
  },
  statusContainer: {
    width: 32,
    marginLeft: 4,
  },
})
