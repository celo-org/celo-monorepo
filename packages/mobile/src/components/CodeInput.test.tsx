import * as React from 'react'
import 'react-native'
import { render } from 'react-native-testing-library'
import CodeInput, { CodeInputStatus } from 'src/components/CodeInput'

describe('CodeInput', () => {
  it('renders correctly for all CodeInputStatus states', () => {
    ;[
      CodeInputStatus.DISABLED,
      CodeInputStatus.INPUTTING,
      CodeInputStatus.PROCESSING,
      CodeInputStatus.RECEIVED,
      CodeInputStatus.ACCEPTED,
    ].map((status) => {
      const { toJSON } = render(
        <CodeInput
          label="label"
          status={status}
          inputValue={'test'}
          inputPlaceholder={'placeholder'}
          onInputChange={jest.fn()}
          shouldShowClipboard={jest.fn()}
        />
      )
      expect(toJSON()).toMatchSnapshot()
    })
  })
})
