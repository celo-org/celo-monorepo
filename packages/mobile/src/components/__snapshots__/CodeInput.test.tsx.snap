// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodeInput renders correctly for all CodeInputStatus states 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#FFFFFF",
        "padding": 16,
      },
      Object {
        "borderRadius": 8,
      },
      undefined,
      Object {
        "backgroundColor": "rgba(103, 99, 86, 0.1)",
        "padding": 0,
      },
    ]
  }
>
  <View
    style={
      Object {
        "borderRadius": 8,
        "overflow": "hidden",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "row",
          "padding": 16,
          "paddingVertical": 12,
        }
      }
    >
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#A49B80",
              "fontFamily": "Inter-SemiBold",
              "fontSize": 13,
              "lineHeight": 16,
              "marginBottom": 4,
              "opacity": 0.5,
            }
          }
        >
          label
        </Text>
        <Text
          numberOfLines={1}
          style={
            Object {
              "color": "#A49B80",
              "fontFamily": "Inter-Regular",
              "fontSize": 16,
              "lineHeight": 22,
            }
          }
        >
          test
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`CodeInput renders correctly for all CodeInputStatus states 2`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#FFFFFF",
        "padding": 16,
      },
      Object {
        "borderRadius": 8,
      },
      Object {
        "elevation": 5,
        "shadowColor": "rgba(48, 46, 37, 0.15)",
        "shadowOffset": Object {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 1,
        "shadowRadius": 8,
      },
      Object {
        "padding": 0,
      },
    ]
  }
>
  <View
    style={
      Object {
        "borderRadius": 8,
        "overflow": "hidden",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "padding": 16,
          "paddingBottom": 4,
        }
      }
    >
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#2E3338",
              "fontFamily": "Inter-SemiBold",
              "fontSize": 13,
              "lineHeight": 16,
            }
          }
        >
          label
        </Text>
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "center",
              },
              undefined,
            ]
          }
        >
          <TextInput
            allowFontScaling={true}
            onBlur={[Function]}
            onChangeText={[MockFunction]}
            onFocus={[Function]}
            placeholder="placeholder"
            rejectResponderTermination={true}
            style={
              Array [
                Object {
                  "color": "#2E3338",
                  "flex": 1,
                  "fontFamily": "Inter-Regular",
                  "fontSize": 16,
                  "lineHeight": 22,
                  "paddingHorizontal": 0,
                  "paddingVertical": 12,
                },
                undefined,
                undefined,
              ]
            }
            underlineColorAndroid="transparent"
            value="test"
          />
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`CodeInput renders correctly for all CodeInputStatus states 3`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#FFFFFF",
        "padding": 16,
      },
      Object {
        "borderRadius": 8,
      },
      undefined,
      Object {
        "backgroundColor": "rgba(103, 99, 86, 0.1)",
        "padding": 0,
      },
    ]
  }
>
  <View
    style={
      Object {
        "borderRadius": 8,
        "overflow": "hidden",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "row",
          "padding": 16,
          "paddingVertical": 12,
        }
      }
    >
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#A49B80",
              "fontFamily": "Inter-SemiBold",
              "fontSize": 13,
              "lineHeight": 16,
              "marginBottom": 4,
              "opacity": 0.5,
            }
          }
        >
          label
        </Text>
        <Text
          numberOfLines={1}
          style={
            Object {
              "color": "#A49B80",
              "fontFamily": "Inter-Regular",
              "fontSize": 16,
              "lineHeight": 22,
            }
          }
        >
          test
        </Text>
      </View>
      <View
        style={
          Object {
            "marginLeft": 4,
            "width": 32,
          }
        }
      >
        <ActivityIndicator
          animating={true}
          color="#1AB775"
          hidesWhenStopped={true}
          size="small"
        />
      </View>
    </View>
  </View>
</View>
`;

exports[`CodeInput renders correctly for all CodeInputStatus states 4`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#FFFFFF",
        "padding": 16,
      },
      Object {
        "borderRadius": 8,
      },
      undefined,
      Object {
        "backgroundColor": "rgba(103, 99, 86, 0.1)",
        "padding": 0,
      },
    ]
  }
>
  <View
    style={
      Object {
        "borderRadius": 8,
        "overflow": "hidden",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "row",
          "padding": 16,
          "paddingVertical": 12,
        }
      }
    >
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#A49B80",
              "fontFamily": "Inter-SemiBold",
              "fontSize": 13,
              "lineHeight": 16,
              "marginBottom": 4,
              "opacity": 0.5,
            }
          }
        >
          label
        </Text>
        <Text
          numberOfLines={1}
          style={
            Object {
              "color": "#A49B80",
              "fontFamily": "Inter-Regular",
              "fontSize": 16,
              "lineHeight": 22,
            }
          }
        >
          test
        </Text>
      </View>
      <View
        style={
          Object {
            "marginLeft": 4,
            "width": 32,
          }
        }
      >
        <ActivityIndicator
          animating={true}
          color="#1AB775"
          hidesWhenStopped={true}
          size="small"
        />
      </View>
    </View>
  </View>
</View>
`;

exports[`CodeInput renders correctly for all CodeInputStatus states 5`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#FFFFFF",
        "padding": 16,
      },
      Object {
        "borderRadius": 8,
      },
      undefined,
      Object {
        "backgroundColor": "rgba(103, 99, 86, 0.1)",
        "padding": 0,
      },
    ]
  }
>
  <View
    style={
      Object {
        "borderRadius": 8,
        "overflow": "hidden",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "row",
          "padding": 16,
          "paddingVertical": 12,
        }
      }
    >
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#A49B80",
              "fontFamily": "Inter-SemiBold",
              "fontSize": 13,
              "lineHeight": 16,
              "marginBottom": 4,
              "opacity": 0.5,
            }
          }
        >
          label
        </Text>
        <Text
          numberOfLines={1}
          style={
            Object {
              "color": "#A49B80",
              "fontFamily": "Inter-Regular",
              "fontSize": 16,
              "lineHeight": 22,
            }
          }
        >
          test
        </Text>
      </View>
      <View
        style={
          Object {
            "marginLeft": 4,
            "width": 32,
          }
        }
      >
        <svg
          fill="none"
          height={32}
          style={Object {}}
          viewBox="0 0 32 32"
          width={32}
        >
          <path
            d="M8.414 15.086a2 2 0 10-2.828 2.828l2.828-2.828zM12.5 22l-1.414 1.414a2 2 0 002.8.029L12.5 22zm13.885-10.557a2 2 0 00-2.77-2.886l2.77 2.886zm-20.8 6.471l5.5 5.5 2.83-2.828-5.5-5.5-2.83 2.828zm8.3 5.529l12.5-12-2.77-2.886-12.5 12 2.77 2.886z"
            fill="#42D689"
            style={Object {}}
          />
        </svg>
      </View>
    </View>
  </View>
</View>
`;
